woocommerce/woocommerce-ios
# [Bundle products in order form] Workaround for optional & non-selected variable bundle item API issue

Closes: #11232 


## Why


As discovered in https://github.com/woocommerce/woocommerce-ios/pull/11186#pullrequestreview-1736581081 - point 1.2:

> If the child item is optional, I can add the bundle to the order but the selected child item(s) don't appear in the order (only the parent bundle item appears).

This is due to the product bundles API issue Pe5pgL-3Vd-p2#api-issues, where an optional & non-selected variable item is supposed to be skipped when `optional_selected` is `false`. However, the `false` boolean is currently causing the variable item to still be validated so that a valid variation ID & attributes are required (which doesn't work for our use cases because we don't always have the variable product at hand like when updating the quantity of the bundle). Before the API is fixed for the majority of the extension versions, we still need a workaround. It's quite hacky, but passing a value that doesn't result in PHP `empty` to be `true` would work. This workaround sets `optional_selected` to be `"no"` instead of `false`. When `optional_selected` is `true`, it remains `true`.

## How

With this workaround, https://github.com/woocommerce/woocommerce-ios/pull/11222 was reverted with updated test cases. The `optional_selected` change is in `OrderItemBundleItem`'s encode function. Then in cases when converting an optional & non-selected variable bundle item, it now returns a bundle configuration with just a product ID without a full set of variation data since it's not selected.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with **1 optional variable item with non-empty variations** and another non-variable item.

- Go to the orders tab
- Tap `+` to create an order
- Tap `Add Products` and select the bundle product in the prerequisite
- Configure the non-variable item if needed so that the configuration is valid and contains at least one child item, then tap `Done` to save the bundle
- Tap `1 Product Selected` to confirm product selection
- Tap `Create` to create the order --> the bundle product should be added with the child item(s) to the order
- Tap `Edit` to edit the order
- Tap to expand the bundle item then increment the quantity to some higher value --> after the order is synced, the quantity for the bundle product and child item(s) should be updated

## Screenshots



https://github.com/woocommerce/woocommerce-ios/assets/1945542/9fed36c3-54bf-4666-bca6-c75e3829d4e3



---
